src/firebolt/common/token_storage.py

```diff
@@ -0,0 +1,114 @@
+import base64
+import json
+import os
+from json import JSONDecodeError
+from typing import Optional
+
+from appdirs import user_data_dir
+from cryptography.fernet import Fernet, InvalidToken
+from cryptography.hazmat.primitives import hashes
+from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC
+
+
+class TokenSecureStorage:
+    def __init__(self, username: str, password: str):
+        """
+        Class for permanent storage of token in the filesystem in encrypted way
+
+        :param username: username used for toke encryption
+        :param password: password used for toke encryption
+        """
+        self._data_dir = user_data_dir(appname="firebolt")
+        os.makedirs(self._data_dir, exist_ok=True)
+
+        self._token_file = os.path.join(self._data_dir, "token.json")
+
+        self.salt = self._get_salt()
+        self.encrypter = FernetEncrypter(self.salt, username, password)
+
+    def _get_salt(self) -> str:
+        """
+        Get salt from the file if exists, or generate a new one
+
+        :return: salt
+        """
+        res = self._read_data_json()
+        if "salt" not in res:
+            return FernetEncrypter.generate_salt()
+        else:
+            return res["salt"]
+
+    def _read_data_json(self) -> dict:
+        """
+        Read json file token.json
+
+        :return: json object as dict
+        """
+        if not os.path.exists(self._token_file):
+            return {}
+
+        with open(self._token_file) as f:
+            try:
+                return json.load(f)
+            except JSONDecodeError:
+                return {}
+
+    def get_cached_token(self) -> Optional[str]:
+        """
+        Get decrypted token using username and password
+        If the token not found or token cannot be decrypted using username, password
+        None will be returned
+
+        :return: token or None
+        """
+        res = self._read_data_json()
+        if "token" not in res:
+            return None
+
+        return self.encrypter.decrypt(res["token"])
+
+    def cache_token(self, token: str) -> None:
```

Comment:
It could be useful to also store token expiration here. This way we can reduce the amount of requests by skipping expired tokens